crates/bdk/src/wallet/mod.rs

```diff
@@ -428,27 +429,48 @@ impl<D> Wallet<D> {
             .next()
     }
 
-    /// Return a single transactions made and received by the wallet
+    /// Calculates the fee of a given transaction. Returns 0 if `tx` is a coinbase transaction.
     ///
-    /// Optionally fill the [`TransactionDetails::transaction`] field with the raw transaction if
-    /// `include_raw` is `true`.
-    pub fn get_tx(&self, txid: Txid, include_raw: bool) -> Option<TransactionDetails> {
+    /// Note `tx` does not have to be in the graph for this to work.
+    pub fn calculate_fee(&self, tx: &Transaction) -> Result<u64, CalculateFeeError> {
+        self.indexed_graph.graph().calculate_fee(tx)
+    }
+
+    /// Calculate the `FeeRate` for a given transaction.
+    ///
+    /// Note `tx` does not have to be in the graph for this to work.
+    pub fn calculate_fee_rate(&self, tx: &Transaction) -> Result<FeeRate, CalculateFeeError> {
+        self.calculate_fee(tx).map(|fee| {
+            let weight = tx.weight();
+            FeeRate::from_wu(fee, weight)
+        })
+    }
+
+    /// Computes total input value going from script pubkeys in the index (sent) and the total output
+    /// value going to script pubkeys in the index (received) in `tx`. For the `sent` to be computed
+    /// correctly, the output being spent must have already been scanned by the index. Calculating
+    /// received just uses the transaction outputs directly, so it will be correct even if it has not
+    /// been scanned.
+    pub fn sent_and_received(&self, tx: &Transaction) -> (u64, u64) {
+        self.indexed_graph.index.sent_and_received(tx)
+    }
+
+    /// Return a single `CanonicalTx` made and received by the wallet or `None` if it doesn't
+    /// exist in the wallet
```

Comment:
```
    /// match canonical_tx.observed_as {
    ///     ChainPosition::Confirmed(anchor) => println!(
    ///         "tx is confirmed at height {}, we know this since {}:{} is in the best chain",
    ///         anchor.confirmation_height,
    ///         anchor.anchor_block.height,
    ///         anchor.anchor_block.hash,
    ///     ),
```

Is this correct? I'm still confused about anchors etc, but from my understanding not all the anchors of a transaction have to be in the best chain. A transaction could have an anchor A1 in the best chain, and an anchor A2 which is not in the best chain - in this case, we know that the tx is confirmed because of A1, but we can't say "tx is confirmed at height h2, we know this since A2 is in the best chain"

Reply:
What I understand this is saying is that if the `canonical_tx.observed_as` is `ChainPosition::Confirmed` then it will contain the anchor this is in the best chain, even if there may be other anchors that aren't in the best chain.  If the `canonical_tx` isn't anchored to the best chain it will have `ChainPosition::Unconfirmed` with the last seen time.  The magic looks like it happens in `TxGraph::try_get_chain_position`.

Reply:
Yes, the `ChainPosition::Confirmed` variant only contains one anchor, which is the anchor that anchors the tx in the best chain.